src/solver/tron.jl

```diff
@@ -0,0 +1,212 @@
+using LinearOperators
+# A trust-region solver for unconstrained optimization with bounds
+
+export tron
+
+function active(x, l, u; ϵlu = 1e-4)
+  if isa(ϵlu, Float64)
+    ϵlu = (u-l)*ϵlu
+  end
+  A = Int[]
+  n = length(x)
+  for i = 1:n
+    if x[i] < l[i] + ϵlu[i]
+      push!(A, i)
+    elseif x[i] > u[i] - ϵlu[i]
+      push!(A, i)
+    end
+  end
+  return A
+end
+
+function tron(nlp :: AbstractNLPModel; bfgs :: Bool=true, μ₀ :: Real=1e-2,
+    μ₁ :: Real=1.0, σ :: Real=1.2, verbose=false, itmax :: Integer=100000,
+    timemax :: Real=60, mem :: Integer=5, atol :: Real=1e-8, rtol :: Real=1e-6)
+  l = nlp.meta.lvar
+  u = nlp.meta.uvar
+  f(x) = obj(nlp, x)
+  g(x) = grad(nlp, x)
+  n = nlp.meta.nvar
+  if bfgs
+    if n <= 5
+      mem = n
+    end
+    H = LBFGSOperator(n, mem, scaling = true)
+  end
```

Comment:
I feel like it's up to the model to decide whether it's using a quasi-Newton Hessian. It would be easy to define a `LBFGSModel` or `LSR1Model` from any `AbstractNLPModel`. I think that's the way to go.